src/k8s/pkg/k8sd/controllers/update_node_configuration.go

```diff
@@ -0,0 +1,110 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/canonical/k8s/pkg/k8sd/types"
+	"github.com/canonical/k8s/pkg/snap"
+	snaputil "github.com/canonical/k8s/pkg/snap/util"
+	"github.com/canonical/k8s/pkg/utils/k8s"
+)
+
+// UpdateNodeConfigurationController asynchronously performs updates of the cluster config.
+// An updates is triggered by sending the patch to the updateCh.
+type UpdateNodeConfigurationController struct {
+	snap         snap.Snap
+	waitReady    func()
+	newK8sClient func() (*k8s.Client, error)
+	// UpdateCh is used to trigger config updates.
+	UpdateCh chan struct{}
+	// ReconciledCh is used to indicate that a reconcilation loop has finished.
+	ReconciledCh chan struct{}
+}
+
+// NewUpdateNodeConfigurationController creates a new controller.
+func NewUpdateNodeConfigurationController(snap snap.Snap, waitReady func(), newK8sClient func() (*k8s.Client, error)) *UpdateNodeConfigurationController {
+	return &UpdateNodeConfigurationController{
+		snap:         snap,
+		waitReady:    waitReady,
+		newK8sClient: newK8sClient,
+		UpdateCh:     make(chan struct{}, 1),
+		ReconciledCh: make(chan struct{}, 1),
+	}
+}
+
+func (c *UpdateNodeConfigurationController) retryNewK8sClient(ctx context.Context) (*k8s.Client, error) {
```

Comment:
nit: This starts to feel a bit duplicated, I also see the point in having it as  part of the controller.

Reply:
ack, let's keep it here for now. If things start to accumulate we can still refactor this.